### ISSUE TITLE
New IP Network function cidrnet(ip_address, netmask) and cidrbitmask(netmask)

### ISSUE THREAD

**Issue by carmineesposito73** [hashicorp/terraform#23905]
### Current Terraform Version
```
Terraform v0.12.19
```

### Use-cases
I have some network data sourced in the format subnet + netmask (e.g. `172.16.0.0 mask 255.255.255.0`) and I need to convert into cidr format (e.g. `172.16.0.0/24`)  

### Proposal
add two new functions:
```
> cidrnet("10.10.10.10", "255.255.255.0")
10.10.10.0/24
> cidrnet("192.168.0.200", "255.255.255.128")
192.168.0.128/25
```
```
> cidrbitmask("255.255.255.0")
24
> cidrbitmask("255.255.192.0")
18
```

**Comment by pkolyvas** [comment]
Hey @carmineesposito73,

I'd love to understand a little more about your use case here and develop a better understanding of the workflow involved. 

While the Terraform Team isn't planning on adding these functions ourselves, we'd welcome a community PR if these functions were of broader interest to our community members.

**Comment by carmineesposito73** [comment]
I'm working on something like that:
I can get information on networks in this format
```
    "gateway" = "172.16.1.1"
    "netmask" = "255.255.255.0"
```
and I would like to use this information to assign an IP address to some resource, but I haven't find a way to use the information above, since cidrhost requires a cidr notation
```
cidrhost("172.16.1.0/24", N)
```

**Comment by lvirbalas** [comment]
This would be generally useful to have.

**Comment by cleeistaken** [comment]
This would be useful.

**Comment by qzx** [comment]
@carmineesposito73 Hi, someone cares <3

@pkolyvas the community delivers

**Comment by apparentlymart** [comment]
Hi @carmineesposito73! Thanks for sharing this use-case, and sorry it's been quiet here for a while.

There have been a few situations like this now, where there are potentially many different ways of encoding essentially the same information and where Terraform has adopted one particular form as "the Terraform way". Other examples include encoding TLS certificates in PEM format, timestamps in RFC 3339 format, time _differences_ in a Terraform-specific notation, etc.

In order to avoid the complexity of having the full matrix of such conversions in the core Terraform language, we've typically expected provider implementers to act as the translation layer between the remote API's conventions and Terraform's conventions. Although we've made occasional exceptions to this rule in exceptional cases where alignment seemed particularly difficult due to impedance mismatches or other design concerns, it seems like this situation is one which could be resolved by designing any Terraform provider that deals with IP address ranges can accept them in Terraform's standard notation -- the CIDR syntax -- and translate back and forth to other notations as needed when working with systems that have disagreeing conventions.

Unless there's a particular technical blocker I'm not seeing where a provider would be unable to do this sort of translation automatically, I'd lean towards saying no to adding these functions, because we typically prefer to deal with system-specific quirks in the providers for those systems; the result is typically more user-friendly (results in a simpler configuration without the need for explicit conversions) and avoids a proliferation of various different global functions which can make it harder to find the specific function needed for a particular situation.

**Comment by carmineesposito73** [comment]
I'll go back to provider :-(
https://github.com/vmware/terraform-provider-vcd/issues/437
https://github.com/vmware/terraform-provider-vcd/issues/595

**Comment by qzx** [comment]
Hi @apparentlymart an thank you for your comment and feedback! However I urge you to have an internal discussion about these functions before completely denying this request. The reasons are multiple and as following. 

1. HashiCorp has already broken this "hashicorp way" for CIDR by introducing the cidrnetmask fuction which retruns a subnet mask in classical dotted decimal. This goes straight counter to the statement that these things should be handled purely in provider


2. HashiCorp claims to value beauty in their products and code, by providing this one directional conversion in the core library, you forced users to create very ugly configurations because you've offered them one way translations and in once the environment and team grows, someone is going to want to translate the other way. At the bare minimum, the function that calculates the CIDR mask size (proposed cidrbitmask) function should also be there for to reverse any damage people can do with the cidrnetmask function.  This goes against one of hashicorps core philosophies

3. HashiCorp has it self claimed not to know what their products really are until they are released into the wild and the use cases are found, this is the community telling you back that the way it's being used, this causes a problem. Users of terraform need these functions for various reasons. This one ticket is based on the VMWare provider, sure. My pull request is based on code I needed to deal with FortiGate provider. In other instances these could be data sources coming from legacy systems that need to be converted into "the hashicorp way of things" in the config. 

4. We get the desire to have the providers take care of these things, that is a lovely thought and wouldn't the world be rosy if everyone just knew what was expected and did that. However it's clear that for larger corporations with huge swathes of resources that need to get implemented into their providers have to use code generation and possibly not very big teams to contribute to FOSS. However, since these are also generally providers of equipment in the very large on-prem enterprise business, this minor inconsistency between "hashicorp wants it this way" and "big enterprise company is trying" comes down hard at the user. Us, the community that uses terraform and is responsible for it's adaption.


FortiNet, VMWare, cisco (especially Enterprise which has no provider today) don't cause terraform adoption, IT professionals do. The fewer of us that care to adopt Terraform despite it lacking internal consistency and parity in it's CIDR functions often get stopped in our tracks by the larger team when they see how many complexities we have to build and maintain on top of having to use all the providers. If there isn't a programmer in the house, then tough luck.

However if HashiCorp wants more adoption then I'd recommend supporting the users in the clash of providers, so that there will be enough of us to be able to tell VMWare and FortiNet and others to put effort into their providers and gain velocity in development there of.


tl;dr
The technical blocker is the existence of the cidrnetmask function already should have an inverse.
The cultural blocker is how ugly this makes terraform management for the admins and the user
The corpo-political blocker is that old school IT professionals in large scale enterprise brownfield environments don't adopt terraform cause they're expected to become programmers. And this is the segment of the market most likely to buy your enterprise products as adoption goes along. This includes my company and our clients. So juuuuust saying.

**Comment by poroping** [comment]
So a real-world example for the ```cidrbitmask``` function. A lot of enterprise networking equipment will still return IP addresses as IP/netmask, or, IP netmask.

At the moment if I want to convert them to CIDR format I need to include a lookup table in my locals as there is no way, at least I can see to do this with the built-in functions.

```
locals {
  input = "10.1.45.1/255.255.255.0"
  netmask2cidr = {
    "0.0.0.0"         = "0"
    "128.0.0.0"       = "1"
    "192.0.0.0"       = "2"
    "224.0.0.0"       = "3"
    "240.0.0.0"       = "4"
    "248.0.0.0"       = "5"
    "252.0.0.0"       = "6"
    "254.0.0.0"       = "7"
    "255.0.0.0"       = "8"
    "255.128.0.0"     = "9"
    "255.192.0.0"     = "10"
    "255.224.0.0"     = "11"
    "255.240.0.0"     = "12"
    "255.248.0.0"     = "13"
    "255.252.0.0"     = "14"
    "255.254.0.0"     = "15"
    "255.255.0.0"     = "16"
    "255.255.128.0"   = "17"
    "255.255.192.0"   = "18"
    "255.255.224.0"   = "19"
    "255.255.240.0"   = "20"
    "255.255.248.0"   = "21"
    "255.255.252.0"   = "22"
    "255.255.254.0"   = "23"
    "255.255.255.0"   = "24"
    "255.255.255.128" = "25"
    "255.255.255.192" = "26"
    "255.255.255.224" = "27"
    "255.255.255.240" = "28"
    "255.255.255.248" = "29"
    "255.255.255.250" = "30"
    "255.255.255.252" = "31"
    "255.255.255.255" = "32"
  }

  cidr2netmask = zipmap(values(local.netmask2cidr), keys(local.netmask2cidr))
}

output "input" {
  value = local.input
}

output "output" {
  value = join("/", [split("/", local.input)[0]] , [local.netmask2cidr[split("/", local.input)[1]]])
}
```
```
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

input = "10.1.45.1/255.255.255.0"
output = "10.1.45.1/24"
```

In a perfect world, yes, all these networking vendors would unify and their APIs would output CIDR format or they would write their providers to do the conversion. But until then, ip/mask is still a very very common format and will need to be worked with.

**Comment by xabinapal** [comment]
While I agree with @apparentlymart on having only one way of doing things and leaving the provider the work of dealing with the quirks of each system, I believe that this idea has many use cases beyond the responsibilities of any provider.

I'm facing an scenario where I need to deploy an application that requires both the subnet and netmask (not accepting CIDRs) of an AWS VPC, also created with Terraform. This application is deployed with the Kubernetes provider, and that value is added via environment variables of a `kubernetes_deployment` resource.

The AWS provider only returns, as expected, the CIDR format of the `aws_subnet` resource. Obviously, it's beyond the scope of the Kubernetes provider to transform it into a subnet + netmask and because of that, right now I need to use a workaround like the one @poroping suggested.

This same requirement could be extended to other providers: deploying the same application with the Helm provider, as an AWS ECS Fargate service, GCP App Engine, Azure...

We already have multiple functions to deal with IPs integrated into Terraform's core. This would allow almost complete manipulation of this kind of data.

If HashiCorp does not want to bloat the core with an increasing number of helper functions, creating a new provider (an official one would be great, so it could easily found by anyone) like the TLS one, with data sources outputting every format and valuable info (IP/CIDR, only IP, only CIDR, subnet, netmask, broadcast address, maximum number of hosts...) could also be an alternative.

**Comment by joakimhellum** [comment]
Also agree with @apparentlymart here and like this idea from @xabinapal:

> If HashiCorp does not want to bloat the core with an increasing number of helper functions, creating a new provider (an official one would be great, so it could easily found by anyone) like the TLS one, with data sources outputting every format and valuable info (IP/CIDR, only IP, only CIDR, subnet, netmask, broadcast address, maximum number of hosts...) could also be an alternative.

We experimented with this previously, here is our use case:

```terraform
data "cidr_network" "example" {
  prefix = "192.168.2.0/24"
}

resource "azurerm_redis_firewall_rule" "example" {
  name                = "example-rule"
  redis_cache_name    = azurerm_redis_cache.example.name
  resource_group_name = azurerm_resource_group.example.name
  start_ip            = data.cidr.network.example.first_ip
  end_ip              = data.cidr.network.example.last_ip
}
```

```terraform
data "cidr_network" "example" {
  ip   = "192.168.2.56"
  mask = "255.255.255.252"
}

resource "azurerm_sql_firewall_rule" "example" {
  name                = "example-rule"
  resource_group_name = azurerm_resource_group.example.name
  server_name         = azurerm_sql_server.example.name
  start_ip_address    = data.cidr.network.example.first_ip
  end_ip_address      = data.cidr.network.example.last_ip
}
```

It feels the Terraform functions needed to replace the `cidr_network` data source in the above example would make Terraform core a bit messy. We are publishing this provider in the Terraform Registry as an experiment.

**Comment by carmineesposito73** [comment]
thanks @joakimhellum

here the result of the data source
```
data "cidr_network" "example" {
  ip   = "192.168.0.60"
  mask = "255.255.255.248"
}

example = {
  broadcast = "192.168.0.63"
  first_ip  = "192.168.0.57"
  id        = "192.168.0.56/29"
  ip        = "192.168.0.60"
  last_ip   = "192.168.0.62"
  mask      = "255.255.255.248"
  mask_bits = 29
  network   = "192.168.0.56/29"
  prefix    = "192.168.0.60/29""
}
```

**Comment by crw** [comment]
Thank you for your continued interest in this issue. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement your own functions! We would love to see this implemented as a provider-defined function. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you are new to provider development, learn how to [create a new provider](https://developer.hashicorp.com/terraform/plugin/framework) with the Terraform Plugin Framework. If you have any questions, please visit the [Terraform Plugin Development category in our official forum](https://discuss.hashicorp.com/c/terraform-providers/tf-plugin-sdk/43). 

We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!
